source/server/admin/listeners_handler.cc

```diff
@@ -16,10 +16,20 @@ ListenersHandler::ListenersHandler(Server::Instance& server) : HandlerContextBas
 Http::Code ListenersHandler::handlerDrainListeners(absl::string_view url, Http::ResponseHeaderMap&,
                                                    Buffer::Instance& response, AdminStream&) {
   const Http::Utility::QueryParams params = Http::Utility::parseQueryString(url);
+
   ListenerManager::StopListenersType stop_listeners_type =
       params.find("inboundonly") != params.end() ? ListenerManager::StopListenersType::InboundOnly
                                                  : ListenerManager::StopListenersType::All;
-  server_.listenerManager().stopListeners(stop_listeners_type);
+
+  const bool graceful = params.find("graceful") != params.end();
+  if (graceful) {
+    server_.drainManager().startDrainSequence([this, stop_listeners_type]() {
```

Comment:
If this is called twice does something bad happen? Does this need to be guarded?

Reply:
Oooh good catch. Added a test, added a guard, left a comment because I'm not sure what the admin response should return if you've called /drain_listeners?graceful twice. Should it silently do nothing? (Since the server's already draining...) Or return an error on the admin call? (If so what error?)